src/util/aws/lambda.js

```diff
@@ -1,90 +1,149 @@
-
-  return lambda.getFunction(params).promise()
+import got from 'got'
+import { AWSEnvironmentVariableNotFound } from '../errors'
+
+/*
+const Function = {
+  FunctionName: 'string',
+  // should be ADT but hey this works
+  Code: {
+    CodeSha256: undefined || 'sha',
+    Zip: undefined || 'binary string',
+    s3: undefined || {
+      bucket: 'string',
+      key: 'string'
+    }
+  },
+  Identifier: {
+   Alias: undefined || 'string',
+   Version: undefined || 'string'
+  }
+  Config: {}
 }
+*/
 
-export async function isFunctionDeployed (FunctionName) {
+export async function getFunction ({ FunctionName, Qualifier }) {
   await loadRegion()
+  const lambda = new AWS.Lambda()
 
-  try {
-    await getFunction({ FunctionName })
-    return true
-  } catch (e) {
-    if (e.code !== 'ResourceNotFoundException') { throw e }
-    return false
+  const func = await lambda.getFunction({ FunctionName, Qualifier }).promise()
+  const isVersion = /[0-9]+/.test(Qualifier)
+  return {
+    FunctionName,
+    Identifier: {
+      Version: func.Configuration.Version,
+      Alias: isVersion ? undefined : Qualifier
+    },
+    Code: { CodeSha256: func.Configuration.CodeSha256 },
+    Config: configStripper(func.Configuration)
   }
 }
 
-export async function putFunction (env, config, ZipFile) {
+export async function createFunction ({ FunctionName, Alias, Code, Config }) {
+  validateConfig(Config)
   await loadRegion()
   const lambda = new AWS.Lambda()
 
-  validateConfig(config)
+  const createFunctionParams = { FunctionName, Publish: true, Code: {}, Config }
+  const createAliasParams = { FunctionName, Name: Alias }
 
-  const FunctionName = config.FunctionName
-  const Publish = true
-
-  try {
-    await getFunction({ FunctionName })
-  } catch (e) {
-    if (e.code !== 'ResourceNotFoundException') { throw e }
-    const params = merge({}, config, { Publish, Code: { ZipFile } })
-    await lambda.createFunction(params).promise()
+  if (Code.s3) {
+    createFunctionParams.Code.S3Bucket = Code.s3.bucket
+    createFunctionParams.Code.S3Key = Code.s3.key
+  } else if (Code.Zip) {
+    createFunctionParams.Code.ZipFile = Code.Zip
+  } else {
+    throw new Error('need to upload code')
   }
 
-  await putEnvironment(env, config)
-  return lambda.updateFunctionCode({ ZipFile, FunctionName, Publish }).promise()
+  const newFunc = await lambda.createFunction(createFunctionParams).promise()
+
+  createAliasParams.FunctionVersion = newFunc.Version
+  await lambda.createAlias(createAliasParams).promise()
+
+  return {
+    FunctionName,
+    Identifier: {
+      Alias,
+      Version: newFunc.Version
+    },
+    Config: configStripper(newFunc),
+    Code: { CodeSha256: newFunc.CodeSha256 }
+  }
 }
 
-export async function putEnvironment (env, config, envVars) {
+// might want to check that the following are equal on these two objects:
+// FunctionName, Alias, rest are mutable
+export async function updateFunction (oldFunction, wantedFunction) {
```

Comment:
`wantedFunction` is a misnomer should instead be something along the lines of `functionChanges` which only contains the different options the user wants to change. Could guard against bad change requests e.g. asking for an `Identifier` or `FunctionName`